packages/firestore/src/core/sync_engine.ts

```diff
@@ -367,30 +368,41 @@ export class SyncEngine implements RemoteSyncer, SharedClientStateSyncer {
    */
   async runTransaction<T>(
     updateFunction: (transaction: Transaction) => Promise<T>,
+    backoff: ExponentialBackoff,
     retries = 5
   ): Promise<T> {
     assert(retries >= 0, 'Got negative number of retries for transaction.');
-    const transaction = this.remoteStore.createTransaction();
-    const userPromise = updateFunction(transaction);
-    if (
-      isNullOrUndefined(userPromise) ||
-      !userPromise.catch ||
-      !userPromise.then
-    ) {
-      return Promise.reject<T>(
-        Error('Transaction callback must return a Promise')
-      );
-    }
-    try {
-      const result = await userPromise;
-      await transaction.commit();
-      return result;
-    } catch (error) {
-      if (retries > 0 && this.isRetryableTransactionError(error)) {
-        return this.runTransaction(updateFunction, retries - 1);
+    let transactionError;
+    for (let i = 0; i < retries; i++) {
+      const backoffDeferred = new Deferred<void>();
+      backoff.backoffAndRun(async () => {
+        backoffDeferred.resolve();
+      });
```

Comment:
This is interesting, but feels like a weird dance to have to do.  I'd be tempted to make backoff.backoffAndRun() return a Promise<> so you could instead do something like:

```
await backoff.backoffAndRun(async () => { });
```
But that's a bit of an odd dance too.  So I'd be tempted to add a new backoff method so you can do:
```
await backoff.awaitBackoff();
```
But now this really highlights the pitfall of this pattern.  By structuring your code like this, it will *not* run on the AsyncQueue which is usually a problem (since it may run intertwined with other SDK code leading to race conditions / inconsistency).  For transactions, this is arguably not a problem since we are already not running them on the AsyncQueue. But TBH I'm unclear about to what degree that's intentional.

Anyway, I guess if we stick with this pattern (see below), I would advocate for backoffAndRun() returning a promise so you don't have to do this Deferred dance.